cpp/models/secir/analyze_result.cpp

```diff
@@ -18,52 +18,72 @@
 * limitations under the License.
 */
 #include "secir/analyze_result.h"
+#include "memilio/math/interpolation.h"
 
 #include <algorithm>
 #include <cassert>
 
 namespace mio
 {
 
-/**
- * TODO: extrapolate first and last point
- */
-TimeSeries<double> interpolate_simulation_result(const TimeSeries<double>& simulation_result)
+TimeSeries<double> interpolate_simulation_result(const TimeSeries<double>& simulation_result, const double abs_tol)
 {
-    assert(simulation_result.get_num_time_points() > 0 && "TimeSeries must not be empty.");
-
-    const auto t0      = simulation_result.get_time(0);
-    const auto tmax    = simulation_result.get_last_time();
-    const auto day0    = static_cast<int>(floor(t0));
-    const auto day_max = static_cast<int>(ceil(tmax));
+    const auto t0 = simulation_result.get_time(0);
+    const auto t_max = simulation_result.get_last_time();
+    // add another day if the first time point is equal to day_0 up to absolute tolerance tol
+    const auto day0 = (t0 - abs_tol < std::ceil(t0) - 1) ? std::floor(t0) : std::ceil(t0);
+    // add another day if the last time point is equal to day_max up to absolute tolerance tol
+    const auto day_max = (t_max + abs_tol > std::floor(t_max) + 1) ? std::ceil(t_max) : std::floor(t_max);
+    
+    // create interpolation_times vector with all days between day0 and day_max
+    std::vector<double> tps(static_cast<int>(day_max) - static_cast<int>(day0) + 1);
+    std::iota(tps.begin(), tps.end(), day0);
+    
+    return interpolate_simulation_result(simulation_result, tps, abs_tol);
+}
 
-    auto day = day0;
+TimeSeries<double> interpolate_simulation_result(const TimeSeries<double>& simulation_result, const std::vector<double>& interpolation_times, const double abs_tol)
+{
+    assert(simulation_result.get_num_time_points() > 0 && "TimeSeries must not be empty.");
+    
+    assert((*std::min_element(interpolation_times.begin(), interpolation_times.end()) + abs_tol >= simulation_result.get_time(0) && *std::max_element(interpolation_times.begin(), interpolation_times.end()) - abs_tol <= simulation_result.get_last_time()) && "Interpolation times have lie between simulation times (up to given tolerance).");
+    unused(abs_tol);
+    
+    if ((int)interpolation_times.size() == 0) {
+        std::cout << "Warning: Vector of interpolation times is empty. Returning empty TimeSeries.";
+    }
+    
     TimeSeries<double> interpolated(simulation_result.get_num_elements());
-    interpolated.reserve(day_max - day0 + 1);
-    interpolated.add_time_point(day, simulation_result.get_value(0));
-    day++;
 
-    //interpolate between pair of time points that lie on either side of each integer day
-    for (int i = 0; i < simulation_result.get_num_time_points() - 1;) {
+    int pointer_interp{};
```

Comment:
I have a suspicion that you chose the an intialization that works by accident. But even if it was intentional, please be explicit about zero initialization:
```suggestion
    int pointer_interp = 0;
```

Reply:
Yes, I will change it.